src/allocator/mod.rs

```diff
@@ -0,0 +1,93 @@
+// use crate::arch;
+extern crate spin;
+
+use crate::*;
+
+use core::mem;
+use core::slice;
+
+use spin::RwLock;
+
+lazy_static! {
+    pub static ref PAGE_ALLOC: RwLock<PageAllocator> = RwLock::new(PageAllocator::new());
+}
+
+#[repr(u8)]
+#[derive(Debug, PartialEq, Copy, Clone)]
+enum UsageFlags {
+    Used = 1,
+    Free = 0,
+}
+
+#[derive(Debug, Default)]
+pub struct PageAllocator {
+    usage: &'static mut [UsageFlags],
+    number_of_pages: usize,
+    first_page: usize,
+}
+
+impl PageAllocator {
+    pub fn new() -> PageAllocator {
+        let heap_start: usize = unsafe { &arch::HEAP_START as *const () } as usize;
+        let heap_end: usize = unsafe { &arch::HEAP_END as *const () } as usize;
```

Comment:
We should definitely add some sort of macro or function for this, I feel like it's quite useful and that we're using it quite often

Reply:
Yes I agree. Macro or function.